psl-core/src/main/java/org/linqs/psl/evaluation/statistics/RankingEvaluator.java

```diff
@@ -289,34 +292,241 @@ public double auroc() {
             double newY = (double)tp / (double)totalPositives;
             double newX = (double)fp / (double)totalNegatives;
 
-            area += 0.5 * (newX - prevX) * (newY + prevY);
+            area += 0.5 * (newX - prevX) * Math.abs(newY - prevY) + (newX - prevX) * newY;
             prevY = newY;
             prevX = newX;
         }
 
         // Add the final piece.
-        area += 0.5 * (1.0 - prevX) * (1.0 + prevY);
+        area += 0.5 * (1.0 - prevX) * Math.abs(1.0 - prevY) + (1.0 - prevX) * 1.0;
 
         return area;
     }
 
+    /**
+     * Returns area mean Average Precision.
+     * Assumes predicted GroundAtoms are hard truth values.
+     * Assumes predicted GroundAtoms are formatted as (query, rank)
+     */
+    public double meanAveragePrecision() {
+        // hashtable to hold the number of ground truth positives for each query
+        Map<String, Integer> totalPositives = new Hashtable<>();
+
+//        System.out.println(truth.toString());
+
+        for (GroundAtom atom : truth) {
+            if (atom.getValue() >= threshold) {
```

Comment:
Instead of having this code block inside of a positive guard, use a control escape inside of a negative guard.

So instead of:
```
if (atom.getValue() >= threshold) {
   ...
}
```

Use:
```
if (atom.getValue() < threshold) {
   continue;
}

...
```

In this case, it is not a big deal.
But this can really add up when when you multiple or larger conditionals.
In general, you want to reduce the number of scopes that a programmer has to keep in their mind at any given time.